### ISSUE TITLE
historical_data throws error with multiple Zaehlwerke/OBIS Codes

### ISSUE THREAD

**Issue by tschoerk** [DarwinsBuddy/WienerNetzeSmartmeter#272]
[log.txt](https://github.com/user-attachments/files/17709291/log.txt)

_Originally posted by @ryanbrown-at in https://github.com/DarwinsBuddy/WienerNetzeSmartmeter/issues/267#issuecomment-2469328272_

The live_sensor, which saves the current meter readings, uses a call to historical_data to get the current meter readings. This function works only with one zaehlwerk, since it has usually only one and we didn't know yet, how to data with multiple zaehlwerke looks like.

Opening a new issue from #267 since ryanbrown-at hat multiple different errors.

**Comment by tschoerk** [comment]
@ryanbrown-at
Can you tell me what wienernetzte assumes as your meter reading for the bezugzaehlpunkt? You find it either on this page https://smartmeter-web.wienernetze.at/verbrauchsdaten/zaehlerdaten with the correct zaehlpunkt selected in the table below under Zählerstand in the line 2024 or on the front page ("Ihr Zählerstand am Date") if the bezug zaehlpunkt is the favourite one (indicated by a star under https://smartmeter-web.wienernetze.at/anlagen-und-vertraege/anlagendaten). Thank you!

**Comment by gotaede** [comment]
I have the same issue and can confirm the correct Zählpunkt is marked as favourite. In my case the second Zählpunkt is from my old apartment and inactive. I assume it is ranked first because the number is smaller than the number of my current Zählpunkt.

**Comment by gotaede** [comment]
@tschoerk I've sent the requested logs via email

**Comment by gotaede** [comment]
I either don't understand your instructions correct or i forgot to copy it in my email: 7.472,565

**Comment by tschoerk** [comment]
Sorry, my bad, you wrote it in the mail already. Just looked at the attachements. Thanks for all the information, I can work with that!

**Comment by tschoerk** [comment]
Ok it seems like there's a differentation between a "normal" Tagstrom zaehlpunkt and a warmepumpe zaehlpunkt. Both are consuming, but one uses obisCode 1.8.0 and the Wärmepumpe uses 1.9.0, which is the sum of 1.8.1 and 1.8.2.

I'll work on a fix.

**Comment by W-M-B** [comment]
as you mentioned OBIS = Zählwerke.
normal private households only have 1 OBIS (kWh/h). 
but there are multiple others like Tagstrom / Nachtstrom = called interruptible Power as well.
then Leistung, Blindstrom etc. 
I will try to look for the offical definition for Austria, so the api is ready to eat all the different options.

**Comment by tschoerk** [comment]
Thanks for those. We currently assume the API returns 1.8.0 for consuming zaehlpunkte (the sum of 1.8.x which the API doesn't send) and 2.8.0 (the sum of 2.8.xm which the API doesn't send) for feeding zaehlpunkt. For example:
```
{
	"zaehlwerke": [
		{
			"obisCode": "1-1:1.8.0",
			"einheit": "WH",
			"messwerte": [
				{
					"messwert": 1000000,
					"zeitVon": "2024-11-09T23:00:00.000Z",
					"zeitBis": "2024-11-10T23:00:00.000Z",
					"qualitaet": "VAL"
				}
			]
		}
	],
	"zaehlpunkt": "AT0010000000000000000"
}
```

Now there's another one which is 1.9.0 (the sum of 1.8.x, which the API does send it this case), which is interesting since they don't list this OBIS-Code, so we could not even prepare for it. Apparently that's something specifically they use for heat pumps (AnlagenType = WAERMEPUMPE).

```
{
	"zaehlwerke": [
		{
			"obisCode": "1-1:1.8.1",
			"einheit": "WH",
			"messwerte": [
				{
					"messwert": 1000000,
					"zeitVon": "2024-11-09T23:00:00.000Z",
					"zeitBis": "2024-11-10T23:00:00.000Z",
					"qualitaet": "VAL"
				}
			]
		},
		{
			"obisCode": "1-1:1.8.2",
			"einheit": "WH",
			"messwerte": [			
				{
					"messwert": 1000000,
					"zeitVon": "2024-11-09T23:00:00.000Z",
					"zeitBis": "2024-11-10T23:00:00.000Z",
					"qualitaet": "VAL"
				}
			]
		},
		{
			"obisCode": "1-1:1.9.0",
			"einheit": "WH",
			"messwerte": [
				{
					"messwert": 2000000,
					"zeitVon": "2024-11-09T23:00:00.000Z",
					"zeitBis": "2024-11-10T23:00:00.000Z",
					"qualitaet": "VAL"
				}
			]
		}
	],
	"zaehlpunkt": "AT0010000000000000000"
}
```

The big questions is, are those other OBIS-Codes interesting to anyone? The API does not return the other codes, but maybe there's a possibility for that and if so, maybe it's beneficial to someone.

**Comment by W-M-B** [comment]
I forgot this stuff already, I looked a bit around,
so it seems that 1-1:1.8.0 Wiener Netze uses for Smartmeter kWh as a standard.
but officially 1-1:1.9.0 is the Wirkarbeit bezug, so the official consumption that shoud be used.
so I would need to correct by assumption and treat
– OBIS-Code 1.6.0 Viertelstündliches Leistungsmaximum (Bezug) in Kilowatt (kW)
– OBIS-Code 1.7.0: Momentanleistung in Kilowatt (kW)
– OBIS-Code 1.8.0: Summe der bezogenen Wirkenergie in Kilowattstunden (kWh)
– OBIS-Code 1.8.x (x=1-4): Bezogene Wirkenergie pro Tarif (1.8.1 Niedertarif, 1.8.2. Hochtarif) in Kilowattstunden (kWh)
**- OBIS-CDE 1.9.0: 
- OBIS-CDE 1.9.x:** 
 most likely also for production by a PV 
2.8.0

Source: I found the formal austrian list of OBIS which differs from the smartmeter one (EB utilitites ist the website that publishes the Standards for data exchange and processes within Austrian Energy Market)
https://www.ebutilities.at/documents/20200304112759_MeterCodes_ConsumptionRecord.pdf
Also we might face some Energiegemeinschaften that use different OBIS 
https://energiegemeinschaften.gv.at/wp-content/uploads/sites/19/2024/05/Verbraucher-Obis-Codes_28.04.2024.png

so in short I guess a list of relevant OBIS needs to be used for data ingestions and the others to be ignored. It is safe to say that multiple obis will happen more and more often.

**Comment by W-M-B** [comment]
I would recommend to prepare for n OBIS values but only take the ones interesting and through away the others.
if we have API users that face issues with a different setup, there is then at least some data to resolve this.
I would ingest 1.9.0 and 1.8.0 
also production 2.9.0 and 2.8.0
and ignore the others

**Comment by DarwinsBuddy** [comment]
Thank you all so much for participating in this fact finding mission :sweat_smile: 

:heart: Thanks for your expertise @W-M-B and for your PR @tschoerk

For me the PR looks perfectly fine. Since it's not a high-prio change I'll wait for feedback on #275 before I release.
